docs/RFCS/20191209_optimistic_ddl.md

```diff
@@ -0,0 +1,650 @@
+# Proposal: Manage DDLs on Sharded Tables by Maximizing Schema Compatibility
+
+- Author(s):    [kennytm](https://github.com/kennytm)
+- Last updated: 2019-12-09
+
+## Abstract
+
+After installing Schema Tracker on DM workers, the DM master can compare schemas from all shards and produce a joined schema compatible with all DMLs produced by all shards. By diffing the joined schema before and after applying an upstream DDL, the DM master can produce the real DDL to be executed downstream.
+
+## Background
+
+DM is a tool for synchronizing upstream MySQL databases into a downstream TiDB cluster via [row-based replication](https://dev.mysql.com/doc/refman/8.0/en/replication-formats.html). Whenever an SQL statement is executed on MySQL, some binlog events will be sent to the DM worker, which then translates them back into SQL and executed on TiDB.
+
+![architecture](../media/dm-architecture.png)
+
+One major feature of DM is [merging manually sharded tables](https://pingcap.com/docs/dev/reference/tools/data-migration/features/shard-merge/) into a single table.
+
+![shard-table-merge](../media/shard-table-merge.png)
+
+The design difficulty on merging sharded tables is handling schema change (DDLs). Currently, DM resolves this with pessimistic replication: if a DDL statement is executed on tbl01, a [barrier](https://en.wikipedia.org/wiki/Barrier_%28computer_science%29) (known as sharding DDL lock) would be set up to pause all binlogs from tbl01 (including this DDL). The same is applied on all other tables, until the same DDL statement is executed on all of tbl01 to tblNN, where the barrier would be unlocked.
+
+This approach ignores the actual action done by the DDL statement, e.g. even if executing ddl_1; ddl_2; is equivalent to ddl_3; ddl_4;, DM cannot recognize this and will not automatically unlock the barrier, and [require human intervention](https://pingcap.com/docs/dev/reference/tools/data-migration/features/manually-handling-sharding-ddl-locks/). This makes the whole feature feel risky and error-prone.
+
+Even strictly following DDL execution, if the user [takes a long time to execute all DDLs](https://docs.google.com/document/d/154r0RGBa5uTWsWV8Q1RmIezamcAdElcLQJ3z_7Vr4nk/edit#), it would create a very long backlog and thus cripples DM’s responsiveness.
```

Comment:
This link to an internal doc.

Reply:
link moved in https://github.com/pingcap/dm/pull/734/commits/2b44322c76187c805f6d081ce32e3a0f213a31ef